### PR TITLE
Support processing custom services

### PR DESCRIPTION
**Pull request by gigxz**
Merge with https://github.com/greenriver/hmis-frontend/pull/297

### Changes
* Add a function `seed_hud_service_form_instances` that will automatically specify Service form instances for all of the HUD services based on project type. (Not funder, yet)
* Update the GQL Service type to represent either a custom or hud service
* Update HmisService initializer to choose whether to create a HUD Service or CustomService based on the CustomServiceType
* Simplify ServiceProcessor to work for both types. Custom data elements are supported as well.
* Change `can_manage_inventory` behavior to exclude management of actual Inventory records. This snuck into this branch, but is unrelated..
* Add custom `esg_funding_service` form

### NOT included
- Ability to specify from instance by funder. We need this in order to accurately  choose when to show the HUD service types.

### CODE REVIEW

File: drivers/hmis/lib/hmis_util/service_types.rb [hunk 1]
```diff
@@ -34,5 +31,111 @@ def self.seed_hud_service_types(data_source_id)
...
+        record_type: 143, # W1 Services Provided – HOPWA
+        project_types: [
+          0, # Emergency Shelter
+          1, # Emergency Shelter
+          2, # Transitional Housing
+          3, # PH - Permanent Supportive Housing
+          6, # Services Only
+          12, # Homelessness Prevention
+        ],
+      },
```
Comment (gigxz) [inline drivers/hmis/lib/hmis_util/service_types.rb]: So this feels kinda wonky, but this is the config for which service types should be available for which project types & funders. I didn't get to implementing the funder piece yet, I'm going to pull that to another PR since we don't currently have support for that in Form Instances.

The seed will create `Instance`s that indicate: "use the `service` form for service type X in project type Y"

File: drivers/hmis/lib/hmis_util/json_forms.rb [hunk 2]
```diff
@@ -69,7 +69,7 @@ def self.apply_fragment(item)
     def self.seed_record_form_definitions
       record_forms.each do |identifier, form_definition|
         role = identifier.upcase.to_sym
-        raise "unrecognized role #{role}" unless Hmis::Form::Definition::FORM_ROLES.key?(role)
+        next unless Hmis::Form::Definition::FORM_ROLES.key?(role)
```
Comment (eanders) [inline drivers/hmis/lib/hmis_util/json_forms.rb:72]: On line 98 you check the same thing but with `role.to_sym` I'm guessing they both should be `to_sym` maybe?

File: drivers/hmis/lib/hmis_util/service_types.rb (hunk 1 above)
Comment (eanders) [reply drivers/hmis/lib/hmis_util/service_types.rb]: I don't have a problem with this setup, except maybe each one should be its own constant and then `HUD_SERVICE_INSTANCE_CONFIG` could just be an array of all the possible situations?  Might make maintenance and re-use easier in the future.  
One question though, are we enforcing that if you choose record type 143, you must apply it to one of the project types in the list?  (I think that's probably good, just checking.)

File: drivers/hmis/lib/hmis_util/service_types.rb [hunk 3]
```diff
@@ -34,5 +31,111 @@ def self.seed_hud_service_types(data_source_id)
...
+          2, # TH
+          6, # services only
+          12, # HP
+        ],
+      },
+      {
+        record_type: 143, # W1 Services Provided – HOPWA
+        project_types: [
+          0, # Emergency Shelter
+          1, # Emergency Shelter
```
Comment (eanders) [inline drivers/hmis/lib/hmis_util/service_types.rb]:
```suggestion
          0, # Emergency Shelter - Entry Exit
          1, # Emergency Shelter - NbN
```

**Review by eanders — APPROVED** [review]
This looks great.  Thanks @gigxz.  So awesome to see so many todos and fixmes getting completed!

File: drivers/hmis/lib/hmis_util/json_forms.rb (hunk 2 above)
Comment (gigxz) [reply drivers/hmis/lib/hmis_util/json_forms.rb:72]: Its already a symbol here

File: drivers/hmis/lib/hmis_util/service_types.rb (hunk 1 above)
Comment (gigxz) [reply drivers/hmis/lib/hmis_util/service_types.rb]: I'm not sure I understand the wording of your question.

This sets up in initial form `Instance`s that make services _available_ for certain project types. So for example, if you open up an enrollment in a Homelessness Prevent project, and try to add a service, the dropdown will include `W1 Services Provided – HOPWA` as an option. The instances need to be further specified by Funder, which I'm putting off to a future PR, since those services should only be available based on the funder by default. Again this is just the initial setup, so it can be changed. It would be nice to have a "form instance manager ui" sooner than later, probably.

Does that make sense?

The idea is that if you make a project be HOPWA-funded, then the HOPWA services will be automatically available in the project without you needing to do anything else. (Obviously, that is dependent on adding the funder dependency to this).

Comment (eanders) [reply drivers/hmis/lib/hmis_util/service_types.rb]: Perfectly clear, thanks!

Comment (gigxz) [reply drivers/hmis/lib/hmis_util/service_types.rb]: Okay. I think I understand your question now, and the answer is no. If you wanted to go in manually and say "Enable HOPWA services for all Rapid Re-Housing Projects" there's nothing stopping you from creating an Instance like that (in the rails console / future instance-manager).
